In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("used_phone_price_prediction_1M.csv")

In [10]:
df

,brand,model,release_year,ram_gb,storage_gb,screen_size_inches,battery_capacity,processor_score,camera_score,os_type,...,body_damage,repair_history,water_damage,city_tier,seller_type,warranty_remaining_months,box_available,charger_available,market_demand_score,resale_price
0,Realme,Realme GT,2025,4,256,6.93,4130,47,100,Android,...,0,0,0,Tier1,Individual,11,1,1,99,47891.46
1,Samsung,Galaxy S21,2024,16,64,6.29,5558,81,67,Android,...,0,0,0,Tier3,Store,8,1,0,64,17663.21
2,Apple,iPhone 11,2022,4,128,5.52,5945,59,67,iOS,...,0,0,0,Tier1,Individual,0,0,1,68,67268.20
3,Google,Pixel 6,2024,16,512,5.99,5767,84,57,Android,...,0,0,1,Tier1,Store,11,1,1,50,12626.26
4,Google,Pixel 7,2021,4,128,6.59,5336,98,44,Android,...,0,0,0,Tier3,Store,24,1,1,66,20662.25
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
999995,Samsung,Galaxy A54,2023,6,128,6.61,6358,99,63,Android,...,0,0,0,Tier3,Store,11,1,1,95,56194.05
999996,Samsung,Galaxy A54,2021,16,512,6.19,6142,63,43,Android,...,0,0,0,Tier3,Store,11,0,1,65,2040.32
999997,Apple,iPhone 15,2022,4,256,5.97,5874,79,49,iOS,...,0,0,0,Tier2,Individual,14,0,1,99,31236.16
999998,Apple,iPhone 13,2023,12,512,6.73,5220,53,56,iOS,...,0,0,0,Tier3,Store,17,1,1,57,16288.20


In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 28 columns):
 #   Column                     Non-Null Count    Dtype  
---  ------                     --------------    -----  
 0   brand                      1000000 non-null  object 
 1   model                      1000000 non-null  object 
 2   release_year               1000000 non-null  int64  
 3   ram_gb                     1000000 non-null  int64  
 4   storage_gb                 1000000 non-null  int64  
 5   screen_size_inches         1000000 non-null  float64
 6   battery_capacity           1000000 non-null  int64  
 7   processor_score            1000000 non-null  int64  
 8   camera_score               1000000 non-null  int64  
 9   os_type                    1000000 non-null  object 
 10  has_5g                     1000000 non-null  int64  
 11  original_price             1000000 non-null  int64  
 12  purchase_year              1000000 non-null  int64  
 13  age_months   

In [4]:
df.isna().sum()

brand                        0
model                        0
release_year                 0
ram_gb                       0
storage_gb                   0
screen_size_inches           0
battery_capacity             0
processor_score              0
camera_score                 0
os_type                      0
has_5g                       0
original_price               0
purchase_year                0
age_months                   0
usage_hours_per_day          0
condition                    0
battery_health               0
screen_cracked               0
body_damage                  0
repair_history               0
water_damage                 0
city_tier                    0
seller_type                  0
warranty_remaining_months    0
box_available                0
charger_available            0
market_demand_score          0
resale_price                 0
dtype: int64

In [5]:
df.duplicated().sum()

np.int64(0)

In [6]:
df.select_dtypes(include='object')

,brand,model,os_type,condition,city_tier,seller_type
0,Realme,Realme GT,Android,Excellent,Tier1,Individual
1,Samsung,Galaxy S21,Android,Fair,Tier3,Store
2,Apple,iPhone 11,iOS,Fair,Tier1,Individual
3,Google,Pixel 6,Android,Good,Tier1,Store
4,Google,Pixel 7,Android,Fair,Tier3,Store
...,...,...,...,...,...,...
999995,Samsung,Galaxy A54,Android,Good,Tier3,Store
999996,Samsung,Galaxy A54,Android,Fair,Tier3,Store
999997,Apple,iPhone 15,iOS,Good,Tier2,Individual
999998,Apple,iPhone 13,iOS,Good,Tier3,Store


In [7]:
df = df.drop("brand",axis=1)

In [8]:
category_cols = ["model","os_type","condition","city_tier","seller_type"]

In [9]:
df = pd.get_dummies(df,columns=
                   category_cols,
                   drop_first = True,
                   dtype = int)

In [10]:
df

,release_year,ram_gb,storage_gb,screen_size_inches,battery_capacity,processor_score,camera_score,has_5g,original_price,purchase_year,...,model_iPhone 13,model_iPhone 14,model_iPhone 15,os_type_iOS,condition_Fair,condition_Good,condition_Poor,city_tier_Tier2,city_tier_Tier3,seller_type_Store
0,2025,4,256,6.93,4130,47,100,1,147337,2025,...,0,0,0,0,0,0,0,0,0,0
1,2024,16,64,6.29,5558,81,67,1,75725,2024,...,0,0,0,0,1,0,0,0,1,1
2,2022,4,128,5.52,5945,59,67,1,134358,2023,...,0,0,0,1,1,0,0,0,0,0
3,2024,16,512,5.99,5767,84,57,1,96807,2024,...,0,0,0,0,0,1,0,0,0,1
4,2021,4,128,6.59,5336,98,44,1,74505,2021,...,0,0,0,0,1,0,0,0,1,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
999995,2023,6,128,6.61,6358,99,63,0,140213,2025,...,0,0,0,0,0,1,0,0,1,1
999996,2021,16,512,6.19,6142,63,43,1,30101,2021,...,0,0,0,0,1,0,0,0,1,1
999997,2022,4,256,5.97,5874,79,49,1,49447,2025,...,0,0,1,1,0,1,0,1,0,0
999998,2023,12,512,6.73,5220,53,56,1,53973,2023,...,1,0,0,1,0,1,0,0,1,1


In [11]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error,r2_score,root_mean_squared_error

In [12]:
x = df.drop("resale_price",axis=1)
y = df["resale_price"]

In [13]:
x_train, x_test, y_train,y_test = train_test_split(
    x,y,test_size = 0.20,  random_state = 42
)

In [14]:
model = RandomForestRegressor(
    n_estimators=100,
    max_depth=20,
    n_jobs=-1,
    random_state=42
)

In [ ]:
model

In [15]:
model.fit(x_train,y_train)

RandomForestRegressor(max_depth=20, n_jobs=-1, random_state=42)

In [16]:
y_pred = model.predict(x_test)

In [17]:
y_pred

array([20201.64642347, 38127.5052324 , 22553.81387808, ...,
       10126.96830287, 35102.85919759, 14536.3970852 ], shape=(200000,))

In [18]:
comparion = pd.DataFrame({
    "Actual" : y_test,
    "prediction" : y_pred
})

In [19]:
comparion

,Actual,prediction
987231,18143.48,20201.646423
79954,44779.54,38127.505232
567130,27081.71,22553.813878
500891,15294.01,21455.543492
55399,17980.52,18217.425117
...,...,...
90245,35041.92,36997.479532
639296,15029.76,12245.222179
311939,12198.55,10126.968303
324459,42174.15,35102.859198


In [21]:
import numpy as np

In [22]:
mse = mean_squared_error(y_test,y_pred)
mae = mean_absolute_error(y_test,y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test,y_pred)
print("MAE :", mae)
print("MSE :", mse)
print("RMSE :", rmse)
print("R2 Score :",r2)

MAE : 2499.901109594952
MSE : 10898356.592649769
RMSE : 3301.2659075951105
R2 Score : 0.9609773377366908


In [24]:
comparion['Error'] = comparion['Actual'] - comparion['prediction']
comparion['Error']

987231   -2058.166423
79954     6652.034768
567130    4527.896122
500891   -6161.533492
55399     -236.905117
             ...     
90245    -1955.559532
639296    2784.537821
311939    2071.581697
324459    7071.290802
390499    -697.387085
Name: Error, Length: 200000, dtype: float64

## Conclusion
The Random Forest Regression model was successfully trained on the 1 million-row used phone price dataset. The model achieved an R² score of 96.10%, indicating that it explains most of the variation in used phone prices. The model achieved an MAE of ₹2,499.90 and an RMSE of ₹3,301.27, showing that the predicted prices are reasonably close to the actual prices. Overall, the model demonstrated strong performance for used phone price prediction.